### ISSUE TITLE
Feature Proposal: Remove titles from thumbnails, add dynamic horizontal sizing

### ISSUE THREAD

**Issue by alex285** [home-sweet-gnome/dash-to-panel#611]
this is really a proposal rather a request, b/c even if that feature was available, im not sure i would be using it; i have to see it in actual use :)

consider the next case that we have a two windows of the same or of a different app, that one consumes more space horizontally, and the other more space vertically (eg a tiled window)

![Screenshot from 2019-03-26 12-50-11](https://user-images.githubusercontent.com/11244308/54991835-72d21400-4fc6-11e9-8257-b70d8cb87fd1.png)

Dash to Panel will currently use the same horizontal space for both previews. that makes some sense b/c less width won't fit the title (or trim it too much to read)

to my use i notice that i rarely pick windows on the titles, but mostly pay attention on the thumbnail 

so the proposal is to be a single option that removes the titles bellow the thumbnails, and also adjust the padding depending window size. in this case the second Tilix preview should had less left/right padding

im a bit lazy to boot on W10 to see how they handle that, but im pretty sure they use dynamic horizontal sizing depending window proportions

**Comment by charlesg99** [comment]
Hey alex285, thank you for the suggestion! Yeah right now we are scaling the thumbnail based on the set preview dimensions, but it might be worth a try to see it scaled the other way around (based on the actual window dimensions). Is Windows doing it much differently (can't remember)? Seems related to #576. Thanks!

**Comment by jderose9** [comment]
Windows will reduce the height as much as it can to fit a preview of all the windows at their original aspect ratio. ie. if there's a single window in the preview and it is more wide than narrow, it will reduce the preview popup height so that there is a thin border all the way around. With multiple windows containing mixed aspect ratio, you would get a larger gap above/below the windows with a wider aspect ratio.

**Comment by charlesg99** [comment]
Hey guys, I started playing with the previews to remove the extra padding (although we can preserve the option to have fixed sizes) and I'm looking for some feedback on the animations I have so far:

![Peek 2019-05-17 01-17](https://user-images.githubusercontent.com/33044886/57904648-a4868f00-7841-11e9-894e-e7187714deb2.gif)


I had to ditch the use of the popup menus for the previews, so there is no input grab with this (which is personally the reason I'm working on this). Sorry about the gif, no cursor for god knows what and it is a bit choppy. What do you think? Thanks!

**Comment by alex285** [comment]
is there somewhere the code we can try it? the only feedback i can tell by watching the gif is "WOW!!!" :)

**Comment by digiberk** [comment]
I love it. That's the way it should be!

**Comment by charlesg99** [comment]
Great, I'll keep at it and push to a branch branch when it is "usable". It's a long weekend here in Canada, so hopefully I'll have some time to work on it, eh?

**Comment by alex285** [comment]
i'll spam this a bit, but to give some additional feedback ..not even my videos get 100 likes :)))

![Screenshot from 2019-05-21 04-27-33](https://user-images.githubusercontent.com/11244308/58061944-01a37d00-7b81-11e9-8254-ab8355a10f1e.png)

**Comment by charlesg99** [comment]
@alex285 Hey awesome! Thanks for sharing and cool video! I'm pretty much done with the new previews, only need to tidy up/connect the settings and I'll be ready to push. Stay tuned :)

**Comment by charlesg99** [comment]
The first version of the new previews is available on branch "custom-window-preview". Here be dragons!

To install:
```
git clone https://github.com/home-sweet-gnome/dash-to-panel.git
cd dash-to-panel
git checkout custom-window-preview
make install
```
Then Alt-F2 -> 'r' to restart gnome-shell or log out/in if on Wayland. 

Feedback and testing will be appreciated, Thanks!

**Comment by HelpMeRuth** [comment]
just to leave my first impressions:
1. the borders inherit the global opacity, making it look kinda weird over white backgrounds.
So i suggest having optional separate styling(opacity / color) for it.
2. Seems like it forces a fixed height even when the previews itself are smaller.
So i suggest making the height dynamic just like how its awesomely done with the width

thats it! thanks for this amazing work :)

**Comment by n1lt0n** [comment]
I love it !!!! thx!!!!!

**Comment by xela92** [comment]
> The first version of the new previews is available on branch "custom-window-preview". Here be dragons!
> 
> To install:
> 
> ```
> git clone https://github.com/home-sweet-gnome/dash-to-panel.git
> cd dash-to-panel
> git checkout custom-window-preview
> make install
> ```
> 
> Then Alt-F2 -> 'r' to restart gnome-shell or log out/in if on Wayland.
> 
> Feedback and testing will be appreciated, Thanks!

Testing it right now, this is the best looking window preview EVER. I reaaaally love it. Very fluid animations, behaviour is perfect. Maybe the only thing I'd change is the close window button background, think a little brighter would be cute.

anyway, it rocks!!! :-D

**Comment by charlesg99** [comment]
Thank you for the feedback! Glad it's working good on your end!

> So i suggest making the height dynamic just like how its awesomely done with the width

I played with that quite a bit and concluded that a single size would be easier to understand as a user setting. Under the hood, the window is scaled up as much as possible within a 16:9 ratio. I had to limit that somehow, otherwise a very large window (having, say, a 10:1 ratio) could've potentially filled your screen sideways. Over the weekend I'll try to see if I could reduce/animate the height of the previews container to fit the height of its taller window.


> So i suggest having optional separate styling(opacity / color) for it.

Yeah I wasn't quite sure about what to do with the colors :) I kinda wanted it to work based on the panel color/opacity and so I'm using that as a base for the preview color variations. Looks cool in front of an empty desktop with a nice background, but I agree that it isn't really what people see most of their day. Maybe we could define a minimum opacity for the previews (like 20% or something), even if the panel is set to be almost transparent?

> Maybe the only thing I'd change is the close window button background, think a little brighter would be cute

I tried to make the close button's background a little brighter instead of darker! I pushed the change to the branch, let me know if you like it better!

Thanks!

**Comment by ddnexus** [comment]
Great improvements overall! Thank you!

The animation of the switch-to-window are super smooth now even when they are on different desktop! That's great!

A few comment/issues:

- The lighter background color for the close booton on the preview may look quite ulgy and it is a noticeable exception to the overall theme used for the OS.
  1. The icon used seems to be the wrong one for the different usage. Instead of picking the regular close icon used in all the window title-bars (as it should be, since now the icon is on a title bar and not on the window itself), it picks the icon used to close the window previews displayed in the overview. That is a big icon in most themes (since it is overlapping to usually big preview windows), while in the little frame of DtP looks really huge.
  2. DtP should not add any different background color just for that icon, since while it might look nice in your specific theme may look really ugly and out of style in most others. Please, "If ain't broken don't fix it!" ;)

- If you preview-focus a file window (e.g. nautilus) and start to type... it will type in the actual window, causing the window to search.. and getting an fatal error in the file app. The user interaction with the preview should not be passed to the actual window.

- Sometimes for some reason that looks random, if you have multiple windows for an app (tried in Nautilus), it refuses to select-focus one of them. That is seldom happening but I will keep an eye on it, trying to find a way to reproduce the behavior.

- The contextual menu for the miniature windows is gone. Is that a need, a choice or just a glitch?

**Comment by ddnexus** [comment]
Another glitch of this branch is that if you use autohide, it will hide the bar even if you are hovering on one of its menu, leaving the menu visible and hiding the bar.

**Comment by HelpMeRuth** [comment]
(white on black for max contrast)
maximized program:
![image](https://user-images.githubusercontent.com/17945932/58367915-73fbc080-7ee5-11e9-80d5-6dec0b24d41c.png)
smaller program:
![image](https://user-images.githubusercontent.com/17945932/58367920-7e1dbf00-7ee5-11e9-8c89-42b7ae082e91.png)
wide program:
![image](https://user-images.githubusercontent.com/17945932/58367951-dbb20b80-7ee5-11e9-87fa-019948f64e27.png)

what about just keeping the 16:9 scaling but at the end you just omit the empty space? i really don like the inconsistent borders :)

about theming, i think limiting the opacity is a good idea(or maybe just make it user configurable). It also might be cool to make it follow the dominant icon color (if enabled), more color is more happy :P

**Comment by ddnexus** [comment]
@HelpMeRuth 

> more color is more happy

unless you have migraines triggered by color brightness and contrast, in which case less color and less contrast is more happy (e.g. [Equilux Theme](https://github.com/ddnexus/equilux-theme))

**Comment by charlesg99** [comment]
Hello, some visual tweaks to the opacity and an option to dynamically reduce the size of the preview containers were added to the branch if you'd like to try. Thanks!

**Comment by ddnexus** [comment]
Thanks for improving the background for the preview bar and removing the exception for the close box!.

Is the " option to dynamically reduce the size of the preview containers " the "Use a fixed size for the window preview"? I don't see any difference: it always adjust the size of the container to the window shape, and I would rather disable that because the scale of the previews windows is variable for each window and it is confusing. (e.g. when you search for a small window, you cannot find it because it has been scaled up to the available size).

I understand that scaling the window may be better for viewing more details, but I rather prefer to have the container fixed (to a screen-size shape) and the window displayed in the original proportion and positioned as it is on the screen. A lot easier to find in my case.  I can identify the position and size better than the content, that is very small and often very similar.

**Comment by HelpMeRuth** [comment]
I'm not seeing any difference between enabled / disabled with "Use a fixed size for the window preview" as well

**Comment by charlesg99** [comment]
Hum weird, this is how it looks like over here (same windows, tried on g-s 3.22, 3.28 and 3.32):

Disabled, so it dynamically resizes:

![Peek 2019-05-27 08-12](https://user-images.githubusercontent.com/33044886/58419154-cf4dc000-8057-11e9-9efc-b9d468f9fb02.gif)

Enabled, so it uses a fixed size:

![Peek 2019-05-27 08-14](https://user-images.githubusercontent.com/33044886/58419195-f1dfd900-8057-11e9-876e-8760f685ee34.gif)

The behavior should be something like this: it scales down the window while respecting a maximum aspect ratio of 16:9, so i.e., if a window has a 18:6 ratio and the option is set to "fixed size", the preview will be centered in the preview's frame, but the height of the frame will remain the same. If the "fixed size" option isn't set, the height of the frame is reduced to fit the preview. Also, when "not fixed", if there are more than one preview in the frame, the taller one defines the height of the frame and the other ones are centered. Hope that makes sense.

**Comment by ddnexus** [comment]
I like the fixed behavior!

Why a hard-coded aspect ratio? For me is fine, but it would be weird for screens with a different aspect ratio. The ideal way should be getting the aspect ratio from the screen automatically, not sure if that is possible or easy tough.

**Comment by HelpMeRuth** [comment]
ahh! my bad didnt notice it because i didnt have long ratio'd windows haha.
the only issue standing for me:
chrome maximized:
![image](https://user-images.githubusercontent.com/17945932/58425932-730c8f80-809b-11e9-83f6-c1e3e34e62d5.png)
discord normal sized
![image](https://user-images.githubusercontent.com/17945932/58426019-af3ff000-809b-11e9-96a6-6c9763e33878.png)
telegram normal sized:
![image](https://user-images.githubusercontent.com/17945932/58426391-9be15480-809c-11e9-86b9-03da65d031e1.png)
g tweaks normal sized:
![image](https://user-images.githubusercontent.com/17945932/58426026-b535d100-809b-11e9-8386-88bdc366ee0f.png)
tilix normal sized:
![image](https://user-images.githubusercontent.com/17945932/58426059-cb439180-809b-11e9-808a-6efb33512e3f.png)
tilix maximized:
![image](https://user-images.githubusercontent.com/17945932/58426152-fd54f380-809b-11e9-9eea-4ed04c6c4ac4.png)

from my small test it seems like non gtk apps work fine(kinda they still have a 1px padding around them when not maximized even tho the padding is set to 0px) but gtk apps have some thick padding around themselfs when not maximized

(note: i tweaked the MIN_MENU_ALPHA to 0.8 within the code which is i think my preferred level)

**Comment by ddnexus** [comment]
@charlesg99 it still does not work here. 

What should happen with a very tall window and a very wide window of the same application with fixed size? I would expect to see two equally sized containers; one with a tall window with a lot of space on its sides and another with a wide window with a lot of space above and below.

**Comment by HelpMeRuth** [comment]
@ddnexus
with "Use a fixed size for the window preview" ENABLED it makes the height always the same (defaults to 240) just like how it was in the beginning(so dynamic width only iirc).
When you DISABLE it makes the height dynamic just like the width so there's supposedly minimal empty space in the preview container but still is limited by the set max height
you can see it in the gif's above and my screenshot from discord

**Comment by HelpMeRuth** [comment]
Disabled:
![image](https://user-images.githubusercontent.com/17945932/58430587-3d22d780-80aa-11e9-9b66-1f7f7a29a5df.png)
Enabled:
![image](https://user-images.githubusercontent.com/17945932/58430601-4ca22080-80aa-11e9-897e-075e1459bfab.png)

**Comment by ddnexus** [comment]
@HelpMeRuth thanks, but what about the example in my previous post? Could you post a screenshot? Than you

**Comment by HelpMeRuth** [comment]
using kid3, qt doesnt seem to have resize limits (stupid :p) 
height test->
DISABLED:
![image](https://user-images.githubusercontent.com/17945932/58431977-79f1cd00-80b0-11e9-9089-cc7e23b17276.png)
ENABLED:
![image](https://user-images.githubusercontent.com/17945932/58432008-b3c2d380-80b0-11e9-84bc-d76ca58be4f3.png)

hmm haha it seems like the toggle only counts for height not width, so it doesnt seem to be how you'd like it yet.

width test ->
DISABLED:
![image](https://user-images.githubusercontent.com/17945932/58432115-1ddb7880-80b1-11e9-833b-7cadad748699.png)

ENABLED:
![image](https://user-images.githubusercontent.com/17945932/58432093-0ac8a880-80b1-11e9-9655-80f058998512.png)

conclusion: "fixed" only counts for height, width is still dynamic (according to this issue's original request)

**Comment by ddnexus** [comment]
Thank you @HelpMeRuth. Now I understand why I don't see any difference!

@charlesg99 what is the point of "fixed" if the width is never fixed? As I said in a previous comment, you want to use "fixed" in order to have the window miniature in the context of the (full-screen-like container), easy to pick/identify because it does not force you to understand the content, but just remember the size/position of the window. If you change the relative window scale and the container shape then it defeats its purpose! "Fixed" should be... "fixed" IMO. :)

Personally, I don't see any advantage on saving horizontal real-estate on horizontal screens, unless you have a zillion windows opened for the same app (and in that case good look spotting it anyway), but with the fixed option working properly, it will not be a problem.

**Comment by HelpMeRuth** [comment]
> but with the fixed option working properly, it will not be a problem.

+1 fixed size should actually mean fixed size, 
what about 2 toggles for dynamic width and height? this should resolve all conflicts in interest

**Comment by charlesg99** [comment]
Sorry guys, I'm still thinking about how to present all the options to the users... The thing about the current "size" value is that it works and is easy to understand if the panel is horizontal (height of the preview) OR vertical (width of the preview). Currently the previews "should" work if/when the option to set the panel vertically is added in the future.

I'm leaning towards offering the following options, which I think would allow all configurations while removing the hard coded aspect ratio (currently defined by the dimensions of the primary monitor):
- Size (numeric value, default 240)
- Aspect ratio X (numeric value, default 16)
- Aspect ratio Y (numeric value, default 9)

Both aspect ratio values could be set to "Fixed" with a switch, So as examples, on a horizontal panel:

The current "Fixed" behavior would be obtained by setting:
Aspect ratio X not fixed
Aspect ratio Y fixed

The current "Not Fixed" behavior would be obtained by setting:
Aspect ratio X not fixed
Aspect ratio Y not fixed

The desired (and undesired as per this original issue ;) "All fixed" behavior would be obtained by setting:
Aspect ratio X fixed
Aspect ratio Y fixed

Does it make sense? Any thoughts? My main concern is presenting this to the users in an easy to understand way, while allowing as much customization as possible.

Thanks!

**Comment by ddnexus** [comment]
With just a couple of field + switch you offer all the possible configurations. That's really the widest customization possible with the minimum input required. Great for the user! I don't know how tricky for you though. :)

**Comment by ddnexus** [comment]
I report this here, because it looks like only this branch is affected.
Commit a3c8da8 broke the persistency of the intellihide when hovering on the system menus of the panel.

**Comment by charlesg99** [comment]
Ha sorry about that @ddnexus, you are right! I didn't correctly understand the intellihide issue you were reporting. This should now be fixed. Thanks!

**Comment by ddnexus** [comment]
Glad to have found a better way to explain the problem. It works! Thank you!

**Comment by ddnexus** [comment]
Just to remember a minor cosmetic glitch/improvement.

The close icon of the preview window is the same icon the window gets in the overview. That was kind of ok with the old preview because it was displayed over the window thumbnail, but now that it is displayed in a bar, it should be switched for the close title bar icon. The overview's icon is bigger in most themes and it looks quite out of proportion (and style) when it appears in a bar.

**Comment by charlesg99** [comment]
The new sizing options were added, but I wonder if it is easy enough to understand. Could you please try playing with those and see if they make sense?

**Comment by ddnexus** [comment]
It looks nice and logical, however it may not behave as one would expect - depending of what one would actually expect of course :).

The fixed width is working as expected and on a horizontal frame the effect is evident, however the fixed height on an horizontal frame is not actually so spectacular, since it's easy obliterated by any other window of the same app with a different proportion. 

Another thing about the fixed-fixed option: the scale and position of the windows inside the frame (that is supposed to represent the screen) is modified anyway. I was expecting to see the actual window/screen proportion not altered, but I am not sure whether that would make sense anymore.

In conclusion, all that settings for one small difference might be a burden for maintenance and we may have asked you too much. I'm fine with anything makes sense to you at this point. We have already more than what we may ask for. 

DtP is really amazing! Thank you!!!

**Comment by HelpMeRuth** [comment]
have you looked into [https://github.com/home-sweet-gnome/dash-to-panel/issues/611#issuecomment-496231698](this bug)?
where gtk apps have a bigger padding around them than non gtk apps, when i set padding to 1px the following happens:
maximized windows:
![image](https://user-images.githubusercontent.com/17945932/59122273-a6340600-895a-11e9-87db-8743046435fc.png)
non gtk normal window:
![image](https://user-images.githubusercontent.com/17945932/59122255-987e8080-895a-11e9-8dbb-2332641e2d91.png)
gtk normal window:
![image](https://user-images.githubusercontent.com/17945932/59122242-8f8daf00-895a-11e9-8e76-6edd75f4b36e.png)

apart from this I have no other problems, been using it ever since the code dropped :) 
maybe a slight annoyance is that windows get unfocused when using the window previews.

**Comment by charlesg99** [comment]
Hey HelpMeRuth, thanks for the feedback! 

> maybe a slight annoyance is that windows get unfocused when using the window previews.

Yeah, I removed that... you can pull/install again. The [idea](https://github.com/home-sweet-gnome/dash-to-panel/issues/611#issuecomment-495842322) came from @ddnexus, but it was annoying me as well. Plus, I don't see why hovering over a preview should prevent normal keyboard input... Hovering over the panel doesn't.

> gtk apps have a bigger padding around them than non gtk apps

It seems this is how mutter renders GTK windows. Not sure why, I haven't dug very deep on that, maybe I'll open an issue. The following red outline is what is considered a window (and is the source of the previews), which explains the added padding. Note this is on a black background...

GTK:
![image](https://user-images.githubusercontent.com/33044886/59140411-c4096700-896a-11e9-9154-6cc69f3bac22.png)

Non GTK:
![image](https://user-images.githubusercontent.com/33044886/59140431-0337b800-896b-11e9-9522-bc429e90a999.png)

**Comment by charlesg99** [comment]
@HelpMeRuth Since your comment, that extra padding was all I could see :) 

I dug a little and it turns out that it was GTK's client side decorations that needed to be removed to have consistent paddings. Should be fixed, please pull/install again to see if that works for you. Thanks!

**Comment by HelpMeRuth** [comment]
:) its working! here's one more thing to see:
![image](https://user-images.githubusercontent.com/17945932/59188687-7c582a80-8b78-11e9-9dc2-ccbabebef46d.png)
![image](https://user-images.githubusercontent.com/17945932/59188726-a14c9d80-8b78-11e9-961c-19e272272cb6.png)

non gtk + non maximized windows have slightly bigger padding :rofl:

**Comment by charlesg99** [comment]
Hey thanks! Yes you are right, Qt windows also have some empty space but they aren't declaring themselves as client decorated (Mutter wise). Should be fixed now!

**Comment by charlesg99** [comment]
Hey, long time coming but the new window previews are now on master :tada: Thanks for all the feedback!

**Comment by alex285** [comment]
you're the best maintainer of all times, in all universes!!!
